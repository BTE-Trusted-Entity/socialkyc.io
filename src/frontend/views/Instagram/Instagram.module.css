.heading {
  composes: heading from '../../components/CredentialFlow/CredentialFlow.module.css';
  background-image: url('../Attester/instagram.svg');
}

.buttonsLine {
  text-align: center;
}

.ctaButton {
  composes: ctaButton from '../../components/CredentialFlow/CredentialFlow.module.css';
  display: inline-flex;
  margin: 0;
}

.profile {
  display: grid;
  grid-template-areas: 'key value' 'key value';
  width: calc(250rem / 16);
  gap: 0.5rem;
  margin-bottom: 2rem;
}
