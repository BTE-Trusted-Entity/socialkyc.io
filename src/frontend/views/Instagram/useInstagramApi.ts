import { useMemo } from 'react';

import { quoteTwitch } from '../../../backend/twitch/quoteTwitchApi';
import { requestAttestationTwitch } from '../../../backend/twitch/requestAttestationTwitchApi';
import { attestTwitch } from '../../../backend/twitch/attestTwitchApi';

import { bindToSession } from '../../utilities/bindToSession';
import { authUrlInstagram } from '../../../backend/instagram/authUrlInstagramApi';
import { confirmInstagram } from '../../../backend/instagram/confirmInstagramApi';

export function useInstagramApi(sessionId: string) {
  return useMemo(() => {
    const sessionBound = bindToSession(sessionId);
    return {
      authUrl: sessionBound(authUrlInstagram),
      confirm: sessionBound(confirmInstagram),
      quote: sessionBound(quoteTwitch),
      requestAttestation: sessionBound(requestAttestationTwitch),
      attest: sessionBound(attestTwitch),
    };
  }, [sessionId]);
}
