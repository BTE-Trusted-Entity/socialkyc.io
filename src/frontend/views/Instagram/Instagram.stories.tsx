import { Meta } from '@storybook/react';

import { action } from '@storybook/addon-actions';

import { InstagramTemplate } from './InstagramTemplate';
import { InstagramProfile } from './Instagram';

export default {
  title: 'Views/Instagram',
  component: InstagramTemplate,
} as Meta;

const actions = {
  handleSignInClick: action('Sign in'),
  handleSubmit: action('submit'),
  handleBackup: action('backup'),
  handleTryAgainClick: action('tryAgain'),
};

const profileMock: InstagramProfile = {
  id: '123456789',
  username: 'john_doe',
  account_type: 'PERSONAL',
};

export function Start(): JSX.Element {
  return <InstagramTemplate status="none" processing={false} {...actions} />;
}

export function UrlReady(): JSX.Element {
  return (
    <InstagramTemplate status="urlReady" processing={false} {...actions} />
  );
}

export function Authorizing(): JSX.Element {
  return (
    <InstagramTemplate status="authorizing" processing={false} {...actions} />
  );
}

export function Authorized(): JSX.Element {
  return (
    <InstagramTemplate
      status="authorized"
      processing={false}
      profile={profileMock}
      {...actions}
    />
  );
}

export function QuoteOpen(): JSX.Element {
  return (
    <InstagramTemplate
      status="authorized"
      processing={true}
      profile={profileMock}
      {...actions}
    />
  );
}

export function Attesting(): JSX.Element {
  return (
    <InstagramTemplate status="attesting" processing={false} {...actions} />
  );
}

export function Ready(): JSX.Element {
  return <InstagramTemplate status="ready" processing={false} {...actions} />;
}

export function UnexpectedError(): JSX.Element {
  return (
    <InstagramTemplate
      status="error"
      processing={false}
      flowError="unknown"
      profile={profileMock}
      {...actions}
    />
  );
}

export function Closed(): JSX.Element {
  return (
    <InstagramTemplate
      status="authorized"
      processing={false}
      flowError="closed"
      profile={profileMock}
      {...actions}
    />
  );
}

export function Unauthorized(): JSX.Element {
  return (
    <InstagramTemplate
      status="error"
      processing={false}
      flowError="unauthorized"
      {...actions}
    />
  );
}
