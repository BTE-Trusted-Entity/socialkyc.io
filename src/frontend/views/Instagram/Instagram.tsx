import { useState, useEffect, useCallback, FormEvent } from 'react';

import { useRouteMatch } from 'react-router-dom';

import { IEncryptedMessage } from '@kiltprotocol/types';

import { Session } from '../../utilities/session';

import { paths } from '../../paths';

import { exceptionToError } from '../../utilities/exceptionToError';

import {
  AttestationStatus,
  FlowError,
  InstagramTemplate,
} from './InstagramTemplate';
import { useInstagramApi } from './useInstagramApi';

export interface InstagramProfile {
  login: string;
  id: string;
}

interface Props {
  session: Session;
}

export function Instagram({ session }: Props): JSX.Element {
  const { code, secret } = (useRouteMatch(paths.instagramAuth)?.params as {
    code?: string;
    secret?: string;
  }) || { code: undefined, secret: undefined };
  console.log(code);

  const initialStatus = code && secret ? 'authorizing' : 'none';

  const [flowError, setFlowError] = useState<FlowError>();
  const [status, setStatus] = useState<AttestationStatus>(initialStatus);
  const [processing, setProcessing] = useState(false);

  const instagramApi = useInstagramApi(session.sessionId);

  const [authUrl, setAuthUrl] = useState('');

  const [profile, setProfile] = useState<InstagramProfile>();

  const [backupMessage, setBackupMessage] = useState<IEncryptedMessage>();

  useEffect(() => {
    if (status !== 'none') {
      return;
    }
    (async () => {
      try {
        const url = await instagramApi.authUrl({});
        setAuthUrl(url);
        setStatus('urlReady');
      } catch (error) {
        console.error(error);
        setStatus('error');
        setFlowError('unknown');
      }
    })();
  }, [status, instagramApi]);

  useEffect(() => {
    if (!code || !secret) {
      return;
    }
    (async () => {
      try {
        setProfile(await instagramApi.confirm({ code, secret }));
        setStatus('authorized');
      } catch {
        setStatus('error');
        setFlowError('unauthorized');
      }
    })();
  }, [instagramApi, code, secret]);

  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();
      setProcessing(true);
      setFlowError(undefined);
      try {
        await session.listen(async (message) => {
          try {
            await instagramApi.requestAttestation({ message });
            setStatus('attesting');
            setProcessing(false);

            setBackupMessage(await instagramApi.attest({}));
            setStatus('ready');
          } catch (exception) {
            const { message } = exceptionToError(exception);
            if (message.includes('closed') || message.includes('rejected')) {
              setFlowError('closed');
            } else {
              console.error(exception);
              setFlowError('unknown');
              setStatus('error');
            }
          }
        });

        const message = await instagramApi.quote({});

        await session.send(message);
      } catch (error) {
        console.error(error);
        setStatus('error');
        setFlowError('unknown');
      } finally {
        setProcessing(false);
      }
    },
    [instagramApi, session],
  );

  const handleBackup = useCallback(async () => {
    try {
      if (!backupMessage) {
        throw new Error('No backup message');
      }
      await session.send(backupMessage);
    } catch (exception) {
      const { message } = exceptionToError(exception);
      if (message.includes('closed')) {
        return; // donâ€™t care that the user has closed the dialog
      }
      setFlowError('unknown');
      console.error(exception);
    }
  }, [backupMessage, session]);

  const handleTryAgainClick = useCallback(() => {
    setStatus('none');
    setFlowError(undefined);
  }, []);

  return (
    <InstagramTemplate
      status={status}
      processing={processing}
      handleSubmit={handleSubmit}
      handleBackup={handleBackup}
      handleTryAgainClick={handleTryAgainClick}
      authUrl={authUrl}
      profile={profile}
      flowError={flowError}
    />
  );
}
