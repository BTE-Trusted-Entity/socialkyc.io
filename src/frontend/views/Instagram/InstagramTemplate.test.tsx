// expect cannot be imported because of https://github.com/testing-library/jest-dom/issues/426
import { describe, it, jest } from '@jest/globals';

import { render } from '../../../testing/testing';

import { InstagramTemplate } from './InstagramTemplate';
import { InstagramProfile } from './Instagram';

jest.useFakeTimers();
jest.setSystemTime(new Date('2022-01-03T12:00:00'));

const actions = {
  handleSignInClick: jest.fn(),
  handleSubmit: jest.fn(),
  handleBackup: jest.fn(),
  handleTryAgainClick: jest.fn(),
};

const profileMock: InstagramProfile = {
  id: '123456789',
  username: 'john_doe',
  account_type: 'PERSONAL',
};

describe('InstagramTemplate', () => {
  it('should match snapshot with status=none', async () => {
    const { container } = render(
      <InstagramTemplate status="none" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=urlReady', async () => {
    const { container } = render(
      <InstagramTemplate status="urlReady" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=authorizing', async () => {
    const { container } = render(
      <InstagramTemplate
        status="authorizing"
        processing={false}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=authorized', async () => {
    const { container } = render(
      <InstagramTemplate
        status="authorized"
        processing={false}
        profile={profileMock}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=attesting', async () => {
    const { container } = render(
      <InstagramTemplate status="attesting" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=ready', async () => {
    const { container } = render(
      <InstagramTemplate status="ready" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with flowerror=unauthorized', async () => {
    const { container } = render(
      <InstagramTemplate
        status="error"
        flowError="unauthorized"
        processing={false}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with flowError=closed', async () => {
    const { container } = render(
      <InstagramTemplate
        status="authorized"
        flowError="closed"
        processing={false}
        profile={profileMock}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with flowError=unknown', async () => {
    const { container } = render(
      <InstagramTemplate
        status="error"
        flowError="unknown"
        processing={false}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
