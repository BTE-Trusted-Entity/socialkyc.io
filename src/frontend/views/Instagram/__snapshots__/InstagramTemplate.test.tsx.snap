// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstagramTemplate should match snapshot with flowError=closed 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Instagram Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <form>
      <dl
        class="profile"
      >
        <dt>
          User-ID:
        </dt>
        <dd>
          123456789
        </dd>
        <dt>
          Username:
        </dt>
        <dd>
          john_doe
        </dd>
        <dt>
          Account Type:
        </dt>
        <dd>
          PERSONAL
        </dd>
      </dl>
      <p>
        Validity: one year (
        1/3/2023
        )
      </p>
      <section
        class="container"
      >
        <div
          class="exclamation"
        />
        <h2
          class="heading"
        >
          Signing error:
        </h2>
        <p
          class="message"
        >
          Your wallet was closed before the request was signed.
        </p>
        <p>
          Click “Continue in Wallet” to try again.
        </p>
      </section>
      <button
        class="ctaButton"
      >
        Continue in Wallet
      </button>
    </form>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`InstagramTemplate should match snapshot with flowError=unknown 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Instagram Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="exclamation"
      />
      <h2
        class="heading"
      >
        Attestation error:
      </h2>
      <p
        class="message"
      >
        Something went wrong!
      </p>
      <p>
        Click “Try Again” or reload the page or restart your browser.
      </p>
    </section>
    <button
      class="ctaButton"
      type="button"
    >
      Try again
    </button>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`InstagramTemplate should match snapshot with flowerror=unauthorized 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Instagram Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="exclamation"
      />
      <h2
        class="heading"
      >
        Authorization error:
      </h2>
      <p
        class="message"
      >
        There was an error authorizing your instagram account.
      </p>
      <p>
        Click “Try Again” or reload the page or restart your browser.
      </p>
    </section>
    <button
      class="ctaButton"
      type="button"
    >
      Try again
    </button>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`InstagramTemplate should match snapshot with status=attesting 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Instagram Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="spinner"
      />
      <h2
        class="heading"
      >
        Attestation process:
      </h2>
      <p
        class="message"
      >
        Anchoring credential on KILT blockchain
      </p>
      <p>
        Please leave this tab open until your credential is attested.
      </p>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`InstagramTemplate should match snapshot with status=authorized 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Instagram Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <form>
      <dl
        class="profile"
      >
        <dt>
          User-ID:
        </dt>
        <dd>
          123456789
        </dd>
        <dt>
          Username:
        </dt>
        <dd>
          john_doe
        </dd>
        <dt>
          Account Type:
        </dt>
        <dd>
          PERSONAL
        </dd>
      </dl>
      <p>
        Validity: one year (
        1/3/2023
        )
      </p>
      <button
        class="ctaButton"
      >
        Continue in Wallet
      </button>
    </form>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`InstagramTemplate should match snapshot with status=authorizing 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Instagram Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="spinner"
      />
      <h2
        class="heading"
      >
        Attestation process:
      </h2>
      <p
        class="message"
      >
        Authorizing
      </p>
      <p>
        Please wait, accessing instagram account details.
      </p>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`InstagramTemplate should match snapshot with status=none 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Instagram Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="spinner"
      />
      <h2
        class="heading"
      >
        Attestation process:
      </h2>
      <p
        class="message"
      >
        Generating link
      </p>
      <p>
        Please wait, generating instagram authorization link.
      </p>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`InstagramTemplate should match snapshot with status=ready 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Instagram Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="checkmark"
      />
      <h2
        class="heading"
      >
        Attestation process:
      </h2>
      <p
        class="message"
      >
        Credential is ready
      </p>
      <p>
        We recommend that you back up your credential now.
      </p>
    </section>
    <button
      class="ctaButton"
      type="button"
    >
      Show credential in wallet
    </button>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`InstagramTemplate should match snapshot with status=urlReady 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Instagram Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <p
      class="buttonsLine"
    >
      <a
        class="ctaButton"
        rel="noreferrer"
        target="_blank"
      >
        Sign in with Instagram
      </a>
    </p>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;
